flink-ml-lib/src/main/java/org/apache/flink/ml/classification/knn/KnnModel.java

```diff
@@ -0,0 +1,223 @@
+            Map<Double, Double> detail = new HashMap<>(0);
+            for (Double obj : tuple2.f0) {
+                detail.merge(obj, percent, Double::sum);
+            }
+            double max = 0.0;
+            double prediction = 0.0;
+            for (Map.Entry<Double, Double> entry : detail.entrySet()) {
+                if (entry.getValue() > max) {
+                    max = entry.getValue();
+                    prediction = entry.getKey();
+                }
+            }
+            return Row.join(row, Row.of(prediction));
+        }
+
+        /** Finds the nearest k neighbors from whole vectors in matrix. */
+        private Tuple2<List<Double>, List<Double>> findNeighbor(DenseVector input) {
+            PriorityQueue<Tuple2<Double, Double>> priorityQueue =
+                    new PriorityQueue<>(Comparator.comparingDouble(o -> -o.f0));
+            double d = 0.0;
+            for (int i = 0; i < input.size(); ++i) {
+                d += input.values[i] * input.values[i];
+            }
+            Tuple2<DenseVector, Double> sample = Tuple2.of(input, d);
+            DenseMatrix packedFeatures = knnModelData.packedFeatures;
+            double[] labelValues = knnModelData.labels.values;
+            double[] normValues = knnModelData.featureNorms.values;
+            DenseVector distanceVector = new DenseVector(packedFeatures.numCols());
+            BLAS.gemv(-2.0, packedFeatures, true, sample.f0, 0.0, distanceVector);
+            for (int i = 0; i < distanceVector.values.length; i++) {
+                distanceVector.values[i] =
+                        Math.sqrt(Math.abs(distanceVector.values[i] + sample.f1 + normValues[i]));
+            }
+            List<Tuple2<Double, Double>> distances = new ArrayList<>(labelValues.length);
+            for (int i = 0; i < labelValues.length; i++) {
+                Tuple2<Double, Double> tuple2 = Tuple2.of(distanceVector.values[i], labelValues[i]);
+                distances.add(tuple2);
+            }
+            Tuple2<Double, Double> head = null;
+            for (Tuple2<Double, Double> currentValue : distances) {
+                if (priorityQueue.size() < k) {
+                    priorityQueue.add(Tuple2.of(currentValue.f0, currentValue.f1));
+                    head = priorityQueue.peek();
+                } else {
+                    if (priorityQueue.comparator().compare(head, currentValue) < 0) {
+                        Tuple2<Double, Double> peek = priorityQueue.poll();
+                        if (peek != null) {
+                            peek.f0 = currentValue.f0;
+                            peek.f1 = currentValue.f1;
+                            priorityQueue.add(peek);
+                            head = priorityQueue.peek();
+                        }
+                    }
+                }
+            }
+            List<Double> items = new ArrayList<>();
+            List<Double> metrics = new ArrayList<>();
```

Comment:
nits: could we make these variable names more consistent with the above code and also more meaningful? Maybe something like `labels` and `distances`.

Reply:
OK